rfc(0006): creating and registering form plugins

ncphillips opened the pull request.

dwalkr commented on 0006-form-hook-conventions.md:
```diff
@@ -0,0 +1,238 @@
+
+I'm not yet convinced this is a geneic problem.
+
+It seems to me that this is only a problem for people using the file system as a storage medium. This includes Git, GitHub, GitLab, etc. For now, I happy to implement this change only at the levels of the `git` and `github` packages. If in the future we realize that this pattern is useful at a more generic level, then it can be introduced to `@tinacms/forms`.
```
agree 100%

jamespohalloran replied:
Agreed, I'd also love to pivot away from filesystem-git being our main example in the docs (since it doesn't scale well on the cloud, or with Next)

jamespohalloran commented:
I know that git as the backend is the one we started with, but I think it's going to prove to be a bigger outlier for approach compared to any other API based data-sources. 
I think I'd probably take it a step further and make just `useForm` the norm, and plugin your own initialData, and onSubmit implementations. For most API backends, it will probably be one-line each.

That way, when you're going through the docs wiring this up on your site, you could get a tina form working much faster, without having to worry about saving until later

dwalkr commented on 0006-form-hook-conventions.md:
````diff
@@ -0,0 +1,238 @@
+```tsx
+const [, form] = useFormatForm(...)
+
+useFormLocally(form)
````
while we're here, is `useFormLocally` still somewhat misleading? I like it a lot better than `useLocalForm` but all this really means is that the form is made accessible from the sidebar (same with `useFormGlobally` and the utility menu.)

I know that the term "local" is meant to denote that the form is removed from the sidebar when navigating to a different page, but that behavior is dependent on where this code is written. You could call `useFormGlobally` and still have a form behave with the intent of a local form, but be placed in the utility menu.

would this be better off as `useFormInSidebar` / `useFormInGlobalMenu` ?

jeffsee55 replied:
I can second this confusion - in reality they're both mounted and unmounted with the component right? This may be a tangent to the RFC but is it possible that global forms need to be registered at the root-level? What are some examples of why this doesn't make sense?

dwalkr replied:
you're exactly right. I think the names global/local refer to how we _intend_ for those respective areas to be used, and not what they actually do. connotation vs. denotation

ncphillips replied:
Totally. The issue is the `global`/`local` terms were sort of intentionally disconnected from the UI. "Local Form" just means "the form as a plugin"; and "global form" just means "the form in a screen plugin".

For example, here's how all those hooks would be aliased:
```ts
const useFormLocally = useFormInSidebar = useFormInToolbar = usePlugins
```

Maybe we should skip the pleasantries and just do:

```ts
usePlugin(form)
```
and

```ts
useFormScreenPlugin(form)
```

ncphillips replied:
Possible change

<img width="663" alt="image" src="https://user-images.githubusercontent.com/824015/79230481-0af1a400-7e3b-11ea-9815-1f120feee420.png">

dwalkr replied:
I think I like that better. `useFormPlugin` is pretty UI-agnostic but dooesn't carry the same collision of connotation that "local" has

dwalkr commented:
> I know that git as the backend is the one we started with, but I think it's going to prove to be a bigger outlier for approach compared to any other API based data-sources.
> I think I'd probably take it a step further and make just `useForm` the norm, and plugin your own initialData, and onSubmit implementations. For most API backends, it will probably be one-line each.
> 
> That way, when you're going through the docs wiring this up on your site, you could get a tina form working much faster, without having to worry about saving until later

@jamespohalloran One thing we haven't really dug into is how to handle authentication (or whether we should) with Tina. Would you envision auth to be handled completely externally to Tina, or would a Form hook possibly need to handle some parts of it / run some code that can't be neatly encapsulated in the current `useForm` api?

jamespohalloran commented:
I'd probably start by having it external to Tina. For example, If you were using the Sanity api plugin in your onSubmit, it might be responsible for attaching your jwt

dwalkr commented on 0006-form-hook-conventions.md:
```diff
@@ -0,0 +1,238 @@
+import { useGitForm } from 'react-tinacms-git';
+import { load, dump } from 'js-yaml';
+
+interface Format<T = any> {
```
is there prior art for the names `parse`/`serialize`? I know these names are semantically correct, I just want them to be as unambiguous as possible.

For example, if the names were instead `parse` and `stringify`, `window.JSON` already satisfies this interface!

ncphillips replied:
final-form uses `parse` and `format`, however I like `format` as the name of the thing rather then the process. `serialize` and `stringify` are both fine by me. I kind of like `stringify`

jamespohalloran commented:
What if instead of a `useGitForm` which wraps` useForm`, the user would do:
```
  const [formData, form] = useForm({
    id: "blog",
    label: "blog",
    fields: [
      {
        name: "markdownBody",
        label: "Home Page Content",
        component: "markdown",
      },
    ],
    // save & commit the file when the "save" button is pressed
    onSubmit(formData) {
        return cms.api.git.commit({
          files: [file.relativePath],
          message: `Commit from Tina: Update ${file.relativePath}`,
        });
	},
  });

    // this would watch for formChanges, and write them to filesystem, 
    // and also load in any initial git content
     useGitForm(form) 
```

This way, I could customize the git message to my liking, I could show an alert on save, etc

dwalkr commented:
I just realized that now that we have `loadInitialValues()` in form config, git forms don't really have side effects https://github.com/tinacms/tinacms/blob/master/packages/next-tinacms-markdown/src/use-markdown-form.ts#L78

jeffsee55 commented:
@jamespohalloran I like this too, I for one might like to `stage` my changes rather than commit them, and then have a global form which does the actual commit.

jamespohalloran commented:
> The useGitForm part feels a bit awkward though

The name might be weird, maybe it's something like `useFilesystemCache`
```
 const useFilesystemCache = (form) =>
 const cms = useCMS()
  const writeToDisk = useCallback(
    ({ values }) => {
      const serialize = getSerialize(form.options.format);
      cms.api.git.writeToDisk({
        fileRelativePath: file.relativePath,
        content: serialize(values),
      });
    },
    [file.fileRelativePath]
  );
  useWatchFormValues(form, writeToDisk);
```

Just a way of extending a form's functionality without having to wrap it

jeffsee55 commented:
I was referring to the idea that if you didn't add that hook then your `onSubmit` would do nothing

ncphillips commented:
I'm  lking where this is going @jamespohalloran. It makes it a bit more explicit/manual in a way that is probably an improvement, especially for our docs. It's definitely an challenging predicament. The current approach makes it super easy for that one use-case: editing markdown files in git. Unfortunately that approach confuses things for people who are looking to use different backends.

There's something to be said for always using the same hook, no matter what the backend is. With this approach any additional behaviour is not composed on top of the form hook, but instead supplied in addition to the form hook. I'm thinking that this approach would make a ton of sense for some `react-tinacms-git` and `react-tinacms-github` packages. 

Of course the major downsides is it takes way more work. I think the prime example is `useRemarkForm` from `gatsby-tinacms-remark`. It does _a lot_ for you. Instead of only calling `useRemarkForm(node)` and having a working form you need to write this out [at least 40 lines of boilerplate](https://github.com/tinacms/tinacms/blob/master/packages/gatsby-tinacms-remark/src/useRemarkForm.tsx#L107-L142). Maybe you cold simplify that, but you're still getting into multiple higher-order-function calls, etc. For example, with `useRemarkForm` you really do want it to _always_ write to disk on change since that's the only way it can way it can work for Gatsby. This makes me a big hesitant to completely throw away abstractions like `useGitForm`. Something about babies and bathwater. 

Regarding what @jeffsee55 said about staging:

> I for one might like to stage my changes rather than commit them, and then have a global form which does the actual commit.

I think this is a good point, but in a similar vain I think we could solve this problem without abandoning that higher-form concept. Instead, we could  make properties like `onSubmit` overideable:

```
useGitForm(file, {
  onSubit(data) {
    return cms.api.git.stage([file.relativePath])
  }
})
```

I think I have two take-away from all this:

1. The docs should be agnostic to the backend. They should be based primarly around examples of `useForm`.
2. We should be more cautious of introducing for higher-order form hooks. They should only be used when are really confident in the assumptions they make. Any other behaviour were less confident in should be implemented only as "Sane Defaults" that can be overridden.

I'm going to let this sit for a while before I make more updates to the RFC so feel free to add any other comments.

dwalkr commented:
I think this is, as you suggested, largely a documentation problem. The framework-agnostic docs focus on `useForm` for this reason - it's the main tool and its config can be deconstructed to explain each part. But our practical examples all use higher-order hooks.

I think it would be good to make a blessed version of Nolan's firebase demo or something similar - a basic react app that implements `useForm` with a basic document DB

I expect that other providers, as well as developers, will create their own higher-order form hooks so I don't want to discourage that outright

ncphillips commented:
I split this RFC into two parts. This PR only contains the Form Creation/Registration bits. A new one for Higher Order Form Hooks (HOFH) is in [RFC 0007](https://github.com/tinacms/rfcs/pull/8/files). In that one I updated it a bit to reflect some of the thinking around reasonable defaults and new abstractions. It makes no statements about _when_ HOFHs should be made. Instead it just talks about _how_ they should be made.

ncphillips commented on 0006-form-hook-conventions.md:
```diff
@@ -0,0 +1,71 @@
+title: Creating and Registering Forms in React
+submitter: ncphillips
+reviewers:
+pull_request: https://github.com/tinacms/rfcs/pull/5
```
```suggestion
pull_request: https://github.com/tinacms/rfcs/pull/5
```

ncphillips commented on 0006-form-hook-conventions.md:
```diff
@@ -0,0 +1,71 @@
+---
+title: Creating and Registering Forms in React
```
```suggestion
title: RFC 0007 – Creating and Registering Forms in React
```
